File: ContractScriptManagement/backend/controllers/auth_controller.go
package controllers

import (
	"github.com/gin-gonic/gin"
	"script-management/config"
	"script-management/models"
	"script-management/utils"
)

type RegisterRequest struct {
	Username string `json:"username" binding:"required"`
	Password string `json:"password" binding:"required"`
	Email    string `json:"email"`
	Phone    string `json:"phone"`
	Nickname string `json:"nickname"`
}

type LoginRequest struct {
	Username string `json:"username" binding:"required"`
	Password string `json:"password" binding:"required"`
}

func Register(c *gin.Context) {
	var req RegisterRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	var existingUser models.User
	if err := config.DB.Where("username = ?", req.Username).First(&existingUser).Error; err == nil {
		utils.BadRequest(c, "用户名已存在")
		return
	}

	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		utils.InternalError(c, "密码加密失败")
		return
	}

	user := models.User{
		Username: req.Username,
		Password: hashedPassword,
		Email:    req.Email,
		Phone:    req.Phone,
		Nickname: req.Nickname,
		Role:     "user",
	}

	if err := config.DB.Create(&user).Error; err != nil {
		utils.InternalError(c, "注册失败")
		return
	}

	token, err := utils.GenerateToken(user.ID, user.Username, user.Role)
	if err != nil {
		utils.InternalError(c, "生成Token失败")
		return
	}

	utils.Success(c, gin.H{
		"token": token,
		"user":  user,
	})
}

func Login(c *gin.Context) {
	var req LoginRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	var user models.User
	if err := config.DB.Where("username = ?", req.Username).First(&user).Error; err != nil {
		utils.BadRequest(c, "用户名或密码错误")
		return
	}

	if !utils.CheckPassword(req.Password, user.Password) {
		utils.BadRequest(c, "用户名或密码错误")
		return
	}

	token, err := utils.GenerateToken(user.ID, user.Username, user.Role)
	if err != nil {
		utils.InternalError(c, "生成Token失败")
		return
	}

	utils.Success(c, gin.H{
		"token": token,
		"user":  user,
	})
}

func GetCurrentUser(c *gin.Context) {
	userID := c.GetUint("user_id")

	var user models.User
	if err := config.DB.First(&user, userID).Error; err != nil {
		utils.NotFound(c, "用户不存在")
		return
	}

	utils.Success(c, user)
}

func UpdateProfile(c *gin.Context) {
	userID := c.GetUint("user_id")

	var user models.User
	if err := config.DB.First(&user, userID).Error; err != nil {
		utils.NotFound(c, "用户不存在")
		return
	}

	var req struct {
		Email    string `json:"email"`
		Phone    string `json:"phone"`
		Nickname string `json:"nickname"`
		Avatar   string `json:"avatar"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	if req.Email != "" {
		user.Email = req.Email
	}
	if req.Phone != "" {
		user.Phone = req.Phone
	}
	if req.Nickname != "" {
		user.Nickname = req.Nickname
	}
	if req.Avatar != "" {
		user.Avatar = req.Avatar
	}

	if err := config.DB.Save(&user).Error; err != nil {
		utils.InternalError(c, "更新失败")
		return
	}

	utils.Success(c, user)
}

File: ClockWorkerCleaning/backend/src/controllers/authController.js
const bcrypt = require('bcryptjs');
const { User, Worker } = require('../models/associations');
const { generateToken } = require('../utils/jwt');
const { success, error } = require('../utils/response');

async function register(req, res, next) {
  try {
    const { phone, password, nickName, role = 'user' } = req.body;

    if (!phone || !password) {
      return res.status(400).json(error('手机号和密码不能为空'));
    }

    const existUser = await User.findOne({ where: { phone } });
    if (existUser) {
      return res.status(400).json(error('手机号已注册'));
    }

    const hashedPassword = bcrypt.hashSync(password, 10);
    const user = await User.create({
      phone,
      password: hashedPassword,
      nickName: nickName || `用户${phone.slice(-4)}`,
      role,
    });

    const token = generateToken({
      id: user.id,
      phone: user.phone,
      role: user.role,
    });

    res.json(success({
      token,
      user: {
        id: user.id,
        phone: user.phone,
        nickName: user.nickName,
        avatar: user.avatar,
        role: user.role,
      },
    }, '注册成功'));
  } catch (err) {
    next(err);
  }
}

async function login(req, res, next) {
  try {
    const { phone, password } = req.body;

    if (!phone || !password) {
      return res.status(400).json(error('手机号和密码不能为空'));
    }

    const user = await User.findOne({ where: { phone } });
    if (!user) {
      return res.status(400).json(error('用户不存在'));
    }

    if (user.status !== 1) {
      return res.status(400).json(error('账号已被禁用'));
    }

    const isValid = bcrypt.compareSync(password, user.password);
    if (!isValid) {
      return res.status(400).json(error('密码错误'));
    }

    const token = generateToken({
      id: user.id,
      phone: user.phone,
      role: user.role,
    });

    let workerInfo = null;
    if (user.role === 'worker') {
      workerInfo = await Worker.findOne({ where: { userId: user.id } });
    }

    res.json(success({
      token,
      user: {
        id: user.id,
        phone: user.phone,
        nickName: user.nickName,
        avatar: user.avatar,
        role: user.role,
        balance: user.balance,
        worker: workerInfo,
      },
    }, '登录成功'));
  } catch (err) {
    next(err);
  }
}

async function getProfile(req, res, next) {
  try {
    const user = await User.findByPk(req.user.id);
    if (!user) {
      return res.status(404).json(error('用户不存在'));
    }

    let workerInfo = null;
    if (user.role === 'worker') {
      workerInfo = await Worker.findOne({ where: { userId: user.id } });
    }

    res.json(success({
      id: user.id,
      phone: user.phone,
      nickName: user.nickName,
      avatar: user.avatar,
      role: user.role,
      balance: user.balance,
      worker: workerInfo,
    }));
  } catch (err) {
    next(err);
  }
}

async function updateProfile(req, res, next) {
  try {
    const { nickName, avatar } = req.body;
    const user = await User.findByPk(req.user.id);
    if (!user) {
      return res.status(404).json(error('用户不存在'));
    }

    if (nickName) user.nickName = nickName;
    if (avatar) user.avatar = avatar;
    await user.save();

    res.json(success(null, '更新成功'));
  } catch (err) {
    next(err);
  }
}

module.exports = {
  register,
  login,
  getProfile,
  updateProfile,
};
